File: backend/services/sample_tier.py
"""
Sample tier classification.

Rules (Feb 2026 — agreed with user):
  • Tiles with any dimension ≤ 200 mm OR size "100x300mm" (small metros,
    mosaics etc.) → FREE_SMALL. The actual tile IS the sample — no cutting
    needed. Free of charge, £2.99 delivery.
  • Tiles with largest dimension ≥ 400 mm AND ≥ 400×800mm equivalent
    (so 600x600, 600x1200, 800x800, 1200x1200, 1200x2400 etc.)
    → eligible for FULL_SIZE paid sample (£5) — a 300×600 mm cut piece
    branded "Full Size Sample". Also still eligible for free cut sample.
  • Everything else → FREE_CUT (standard 10×10 cm cut piece). Free,
    £2.99 delivery.

Every tile is always eligible for at least one free tier; the full-size
tier is additive on large-format tiles only.
"""
from __future__ import annotations

from typing import Optional

# Tuning knobs — deliberately conservative so we don't auto-suppress
# the free cut sample for tiles that fitters still want to see cut.
SMALL_TILE_MAX_MM = 200              # any-dimension ≤ this → sample IS the tile
SMALL_TILE_RECT_NARROW_MM = 100      # matches 100×300, 100×400 etc.
SMALL_TILE_RECT_LONG_MM = 300
LARGE_FORMAT_MIN_LONG_MM = 600       # tiles ≥ this long dimension offer full-size paid
LARGE_FORMAT_MIN_SHORT_MM = 400      # AND short dimension ≥ this


def _norm(value: Optional[float]) -> Optional[float]:
    """Accept None / 0 / floats in mm. Returns float mm or None."""
    if value is None:
        return None
    try:
        v = float(value)
    except (TypeError, ValueError):
        return None
    if v <= 0:
        return None
    return v


def classify(
    tile_width: Optional[float],
    tile_height: Optional[float],
) -> dict:
    """Returns tier descriptor for a tile's dimensions.

    Output:
      {
        "primary": "free_small" | "free_cut",   # cheapest/default tier
        "offers": ["free_small"] | ["free_cut"] | ["free_cut", "full_size"],
        "width_mm": float | None,
        "height_mm": float | None,
        "full_size_note": None | str,           # shown on PDP when full-size picked
      }

    If dimensions are unknown → defaults to ["free_cut"] (safest).
    """
    w = _norm(tile_width)
    h = _norm(tile_height)

    if w is None or h is None:
        return {
            "primary": "free_cut",
            "offers": ["free_cut"],
            "width_mm": w,
            "height_mm": h,
            "full_size_note": None,
        }

    long_side = max(w, h)
    short_side = min(w, h)

    # Tier 1 — small tile, the tile IS the sample
    is_small_square = long_side <= SMALL_TILE_MAX_MM
    is_narrow_rect = (
        short_side <= SMALL_TILE_RECT_NARROW_MM
        and long_side <= SMALL_TILE_RECT_LONG_MM
    )
    if is_small_square or is_narrow_rect:
        return {
            "primary": "free_small",
            "offers": ["free_small"],
            "width_mm": w,
            "height_mm": h,
            "full_size_note": None,
        }

    # Tier 3 — large format: additive full-size paid option
    if long_side >= LARGE_FORMAT_MIN_LONG_MM and short_side >= LARGE_FORMAT_MIN_SHORT_MM:
        # Exact 300×600 tiles don't need the "cut from larger" caveat
        if long_side <= 600 and short_side <= 300:
            note = None
        else:
            note = (
                f"This is a large sample cut to approximately 300×600 mm "
                f"from a {int(long_side)}×{int(short_side)} mm tile."
            )
        return {
            "primary": "free_cut",
            "offers": ["free_cut", "full_size"],
            "width_mm": w,
            "height_mm": h,
            "full_size_note": note,
        }

    # Tier 2 — everything else: free cut only
    return {
        "primary": "free_cut",
        "offers": ["free_cut"],
        "width_mm": w,
        "height_mm": h,
        "full_size_note": None,
    }


# Prices — kept here so the backend is the source of truth. Frontend
# reads them via /api/content/sample-service if customised; otherwise
# these defaults apply.
FREE_SAMPLE_POSTAGE_GBP = 2.99
FULL_SIZE_SAMPLE_PRICE_GBP = 5.00

File: frontend/src/lib/sampleTier.js
/**
 * Sample-tier classification — JS mirror of backend/services/sample_tier.py
 *
 * MUST stay in sync with the Python version. Any tuning knob change
 * should happen in both files.
 *
 * Tiers:
 *   • free_small  — tile ≤ 200×200mm or ≤100×300mm. Tile IS the sample.
 *                   FREE + £2.99 delivery.
 *   • free_cut    — standard ~10×10cm cut piece. FREE + £2.99 delivery.
 *   • full_size   — 300×600mm cut piece, branded "Full Size Sample".
 *                   £5 each + £2.99 delivery. Additive on large-format tiles
 *                   (≥400mm short side AND ≥600mm long side).
 */

const SMALL_TILE_MAX_MM = 200;
const SMALL_TILE_RECT_NARROW_MM = 100;
const SMALL_TILE_RECT_LONG_MM = 300;
const LARGE_FORMAT_MIN_LONG_MM = 600;
const LARGE_FORMAT_MIN_SHORT_MM = 400;

export const FREE_SAMPLE_POSTAGE_GBP = 2.99;
export const FULL_SIZE_SAMPLE_PRICE_GBP = 5.00;

function normaliseDim(value) {
  if (value === null || value === undefined) return null;
  const v = parseFloat(value);
  if (!isFinite(v) || v <= 0) return null;
  return v;
}

/**
 * Classify a tile's dimensions into sample tiers.
 *
 * @param {number|string|null} tileWidth  — millimetres
 * @param {number|string|null} tileHeight — millimetres
 * @returns {{
 *   primary: 'free_small' | 'free_cut',
 *   offers: Array<'free_small' | 'free_cut' | 'full_size'>,
 *   widthMm: number|null,
 *   heightMm: number|null,
 *   fullSizeNote: string|null,
 * }}
 */
export function classifySample(tileWidth, tileHeight) {
  const w = normaliseDim(tileWidth);
  const h = normaliseDim(tileHeight);

  if (w === null || h === null) {
    return {
      primary: 'free_cut',
      offers: ['free_cut'],
      widthMm: w,
      heightMm: h,
      fullSizeNote: null,
    };
  }

  const longSide = Math.max(w, h);
  const shortSide = Math.min(w, h);

  const isSmallSquare = longSide <= SMALL_TILE_MAX_MM;
  const isNarrowRect =
    shortSide <= SMALL_TILE_RECT_NARROW_MM &&
    longSide <= SMALL_TILE_RECT_LONG_MM;

  if (isSmallSquare || isNarrowRect) {
    return {
      primary: 'free_small',
      offers: ['free_small'],
      widthMm: w,
      heightMm: h,
      fullSizeNote: null,
    };
  }

  if (longSide >= LARGE_FORMAT_MIN_LONG_MM && shortSide >= LARGE_FORMAT_MIN_SHORT_MM) {
    const note =
      longSide <= 600 && shortSide <= 300
        ? null
        : `This is a large sample cut to approximately 300×600 mm from a ${Math.round(longSide)}×${Math.round(shortSide)} mm tile.`;
    return {
      primary: 'free_cut',
      offers: ['free_cut', 'full_size'],
      widthMm: w,
      heightMm: h,
      fullSizeNote: note,
    };
  }

  return {
    primary: 'free_cut',
    offers: ['free_cut'],
    widthMm: w,
    heightMm: h,
    fullSizeNote: null,
  };
}

/**
 * Extract tile_width / tile_height from a product object. Products have
 * inconsistent field names across the codebase (tile_width, size string,
 * etc.). This helper hides that detail from callers.
 */
export function getTileDims(product) {
  if (!product) return { width: null, height: null };
  // Preferred: explicit numeric fields
  if (product.tile_width && product.tile_height) {
    return { width: product.tile_width, height: product.tile_height };
  }
  // Fallback: parse "600x1200", "600 x 1200mm", "600X1200" from size string
  const sizeStr = product.size || product.tile_size;
  if (typeof sizeStr === 'string') {
    const m = sizeStr.match(/(\d+)\s*[xX×]\s*(\d+)/);
    if (m) return { width: parseInt(m[1], 10), height: parseInt(m[2], 10) };
  }
  return { width: null, height: null };
}

/**
 * Convenience: classify straight from a product object.
 */
export function classifySampleForProduct(product) {
  const { width, height } = getTileDims(product);
  return classifySample(width, height);
}

export const SAMPLE_TIER_LABELS = {
  free_small: 'Free Sample (actual tile)',
  free_cut:   'Free Sample (10×10 cm cut piece)',
  full_size:  'Full Size Sample (300×600 mm) — £5',
};

export const SAMPLE_TIER_SHORT = {
  free_small: 'Free sample',
  free_cut:   'Free sample',
  full_size:  'Full Size £5',
};
